File: SeaBattle/SeaBattle.CSharp/Ship.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace SeatBattle.CSharp
{
    [DebuggerDisplay("({Location.X},{Location.Y}) {Orientation} x{Length}")]
    public class Ship
    {
        public Ship(int length)
        {
            Length = length;
        }

        public int Length { get; set; }
        public ShipOrientation Orientation { get; set; }
        public int X { get; set; }
        public int Y { get; set; }
        public int HitCount { get; set; }

        public bool IsDrowned
        {
            get
            {
                return HitCount == Length;
            }
        }

        public bool IsLocatedAt(int x, int y)
        {
            var rect = GetShipRegion();

            return (x >= rect.X && x <= rect.Right && y >= rect.Y && y <= rect.Bottom);
        }

        public Rect GetShipRegion()
        {
            var width = Orientation == ShipOrientation.Horizontal ? Length : 1;
            var height = Orientation == ShipOrientation.Vertical ? Length : 1;

            return new Rect(X, Y, width, height);
        }

        public bool IsInRegion(Rect rect)
        {
            var r = GetShipRegion();
            return (rect.IntersectsWith(r));
        }

        public void MoveTo(int x, int y)
        {
            X = x;
            Y = y;
        }

        public void Rotate()
        {
            Orientation = Orientation == ShipOrientation.Horizontal ? ShipOrientation.Vertical : ShipOrientation.Horizontal;
        }

    }
}

File: SeaBattle/SeaBattle.Cpp/Ship.cpp
#include "stdafx.h"
#include "Ship.h"
#include "Rect.h"
#include "Enums.h"

Ship::Ship(int length)
{
    Ship(length, ShipOrientation::Horizontal);
};

Ship::Ship(int length, ShipOrientation orientation)
{
    Length = length;
    Orientation = orientation;
};

bool Ship::IsDrowned::get()
{
	return HitCount == Length;
};

bool Ship::IsLocatedAt(int x, int y)
{
	Rect^ rect = GetShipRegion();
    return (x >= rect->X 
            && 
            x <= rect->Right 
            && 
            y >= rect->Y 
            && 
            y <= rect->Bottom);
};

Rect^ Ship::GetShipRegion()
{
	int width = Orientation == ShipOrientation::Horizontal ? Length : 1;
	int height = Orientation == ShipOrientation::Vertical ? Length : 1;

	return gcnew Rect(X, Y, width, height);
};

bool Ship::IsInRegion(Rect^ rect)
{
	Rect^ r = GetShipRegion();
	return (rect->IntersectsWith(r));
}

void Ship::MoveTo(int x, int y)
{
	X = x;
	Y = y;
}

void Ship::Rotate()
{
	Orientation = Orientation == ShipOrientation::Horizontal ? ShipOrientation::Vertical : ShipOrientation::Horizontal;
}